scripts/linters/pylint_extensions.py

```diff
@@ -1654,99 +1654,112 @@ class SingleLineCommentChecker(checkers.BaseChecker):
+        # Comments must end with the proper punctuation.
+        last_char_is_invalid = line[-1] not in (
+            ALLOWED_TERMINATING_PUNCTUATIONS)
+
+        excluded_phrase_at_beginning_of_line = any(
+            line[1:].startswith(word) for word in EXCLUDED_PHRASES)
+        if (last_char_is_invalid and not (
+                excluded_phrase_is_present_at_end or
+                excluded_phrase_at_beginning_of_line)):
+            self.add_message('invalid-punctuation-used', line=line_num)
 
-            if not next_line.startswith(b'#'):
-                # Comments must end with the proper punctuation.
-                last_char_is_invalid = line[-1] not in (
-                    ALLOWED_TERMINATING_PUNCTUATIONS)
-                if last_char_is_invalid:
-                    self.add_message(
-                        'invalid-punctuation-used', line=line_num + 1)
+    def process_tokens(self, tokens):
+        """Custom pylint checker to ensure that comments follow correct style.
 
-            # Comments must start with a space.
-            if re.search(br'^#[^\s].*$', line) and not line.startswith(b'#!'):
-                space_at_beginning_of_comment = False
-                self.add_message(
-                    'no-space-at-beginning', line=line_num + 1)
-
-            # Comments may include a lowercase character at the beginning
-            # only if they start with version info or a data type or.
-            # a variable name e.g. "# next_line is of string type."
-            # or "# v2 version does not have ExplorationStatsModel."
-            # or "# int. The file size, in bytes.".
-            if len(line) > 1 and space_at_beginning_of_comment:
-
-                # Check if variable name is used.
-                underscore_is_present = '_' in line.split()[1]
-                if underscore_is_present:
-                    continue
+        Args:
+            tokens: Token. Object to access all tokens of a module.
+        """
+        prev_line_num = -1
+        comments = []
+        comments_index = -1
 
-                # Check if allowed prefix is used.
-                allowed_prefix_is_present = any(
-                    line[2:].startswith(word) for word in (
-                        allowed_comment_prefixes))
-                if allowed_prefix_is_present:
-                    continue
+        for (token_type, _, (line_num, _), _, line) in tokens:
+            if token_type == tokenize.COMMENT:
+                line = line.strip()
 
-            # Comments must start with a capital letter.
-            if not previous_line.startswith(b'#') and (
-                    re.search(br'^# [a-z][A-Za-z]*.*$', line)):
-                self.add_message(
-                    'no-capital-letter-at-beginning', line=line_num + 1)
+                if line.startswith('#'):
+                    if prev_line_num + 1 == line_num:
+                        comments[comments_index].append((line_num, line))
```

Comment:
I appended line_num and line here instead of token because token is not available at this point, see line 1734.

Reply:
Line 1734 is an empty line! Maybe you can try explaining it? (Though correct line number will also be very helpful)

Reply:
Sorry, I mistyped, the line was 1742.